### PR TITLE
[R-package] [docs] Simplified examples to cut example run time (fixes #2988)

### LINKED ISSUE #2988: [R-package] CRAN note about example timings > 5 seconds
See https://github.com/microsoft/LightGBM/pull/2985#issuecomment-611582338.

The R package has been just barely under this timing I guess:

```
* checking examples ... NOTE
Examples with CPU or elapsed time > 5s
     user system elapsed
dim 5.426  0.264    5.79
```

We hopefully can get back under it by making some of the examples less costly to run or using  `\dontrun`  guards. I would prefer to make the examples simpler, since `\dontrun` guards means that we lose testing of the documentation.

I ran the following on my Mac to get timings for the examples

```shell
Rscript build_r.R --skip-install
R CMD CHECK lightgbm*.tar.gz --as-cran --timings
Rscript -e "
    timing_df <- read.delim('lightgbm.Rcheck/lightgbm-Ex.timings');
    print(timing_df);
    print(paste0('total time (s) ', sum(timing_df[['name']])))
    "
```

The timings are as shown:

```
                             name  user system elapsed
dim                         0.695 0.036  0.731      NA
dimnames.lgb.Dataset        0.056 0.007  0.053      NA
getinfo                     0.060 0.002  0.056      NA
lgb.Dataset                 0.049 0.007  0.049      NA
lgb.Dataset.construct       0.043 0.001  0.035      NA
lgb.Dataset.create.valid    0.034 0.002  0.036      NA
lgb.Dataset.save            0.043 0.002  0.036      NA
lgb.Dataset.set.categorical 0.044 0.002  0.037      NA
lgb.Dataset.set.reference   0.034 0.001  0.036      NA
lgb.cv                      0.284 0.084  0.188      NA
lgb.dump                    0.070 0.012  0.059      NA
lgb.get.eval.result         0.066 0.013  0.060      NA
lgb.importance              0.264 0.042  0.214      NA
lgb.interprete              0.951 0.048  0.507      NA
lgb.load                    0.148 0.021  0.079      NA
lgb.model.dt.tree           0.118 0.029  0.116      NA
lgb.plot.importance         0.130 0.028  0.124      NA
lgb.plot.interpretation     0.926 0.042  0.479      NA
lgb.prepare                 0.020 0.001  0.012      NA
lgb.prepare2                0.010 0.001  0.006      NA
lgb.prepare_rules           0.015 0.002  0.019      NA
lgb.prepare_rules2          0.011 0.002  0.014      NA
lgb.save                    0.064 0.014  0.061      NA
lgb.train                   0.064 0.012  0.060      NA
lgb.unloader                0.066 0.013  0.063      NA
predict.lgb.Booster         0.068 0.013  0.060      NA
readRDS.lgb.Booster         0.076 0.013  0.077      NA
saveRDS.lgb.Booster         0.073 0.014  0.067      NA
setinfo                     0.052 0.003  0.046      NA
slice                       0.052 0.002  0.045      NA
[1] "total time (s) 4.586"
```

it looks like the example for `lgb.interprete()` and `lgb.plot.interpretationo()` take almost 2 seconds to run. I'll see if I can speed them up or just `\dontrun` them.

### PR DESCRIPTION
**Pull request by jameslamb**
See #2988 . The R package is very close to hitting this `R CMD CHECK` NOTE:

```
* checking examples ... NOTE
Examples with CPU or elapsed time > 5s
     user system elapsed
dim 5.426  0.264    5.79
```

In this PR, I cut the runtime for our examples by:

* removing all `library(lightgbm)` (these are not necessary and it is conventional to omit them)
* cutting `nrounds` from  `10L` to smaller numbers in all examples with training
* wrapping examples for code that relies on the filesystem in `\donttest`
* wrapping plotting code in `\donttest`

These tests run in about 2.9 seconds on my laptop. To get the timings for yourself

```shell
pushd R-package
    Rscript -e "roxygen2::roxygenize(load = 'installed')"
popd

Rscript build_r.R --skip-install

R CMD CHECK lightgbm*.tar.gz \
    --no-tests \
    --no-manual \
    --timings

Rscript -e "
    timing_df <- read.delim('lightgbm.Rcheck/lightgbm-Ex.timings');
    print(timing_df);
    print(paste0('total time (s) ', sum(timing_df[['name']])))
    "
```

Results:

```text
                             name  user system elapsed
dim                         0.705 0.038  0.746      NA
dimnames.lgb.Dataset        0.049 0.006  0.046      NA
getinfo                     0.061 0.004  0.057      NA
lgb.Dataset                 0.050 0.008  0.050      NA
lgb.Dataset.construct       0.044 0.002  0.037      NA
lgb.Dataset.create.valid    0.035 0.003  0.037      NA
lgb.Dataset.save            0.042 0.003  0.036      NA
lgb.Dataset.set.categorical 0.043 0.002  0.036      NA
lgb.Dataset.set.reference   0.034 0.003  0.038      NA
lgb.cv                      0.287 0.078  0.203      NA
lgb.dump                    0.001 0.000  0.000      NA
lgb.get.eval.result         0.118 0.017  0.063      NA
lgb.importance              0.167 0.018  0.137      NA
lgb.interprete              0.409 0.024  0.216      NA
lgb.load                    0.000 0.000  0.000      NA
lgb.model.dt.tree           0.251 0.040  0.145      NA
lgb.plot.importance         0.146 0.021  0.108      NA
lgb.plot.interpretation     0.000 0.000  0.000      NA
lgb.prepare                 0.017 0.001  0.009      NA
lgb.prepare2                0.011 0.001  0.006      NA
lgb.prepare_rules           0.028 0.002  0.015      NA
lgb.prepare_rules2          0.023 0.002  0.013      NA
lgb.save                    0.000 0.000  0.000      NA
lgb.train                   0.119 0.017  0.066      NA
lgb.unloader                0.117 0.017  0.062      NA
predict.lgb.Booster         0.081 0.015  0.063      NA
readRDS.lgb.Booster         0.000 0.000  0.000      NA
saveRDS.lgb.Booster         0.000 0.000  0.000      NA
setinfo                     0.052 0.002  0.046      NA
slice                       0.054 0.003  0.048      NA
[1] "total time (s) 2.944"
```

To answer some questions I anticipate getting:

**Should we really be using `\donttest?` That means CI won't catch those examples being broken**

I think it's worth it to get rid of the `R CMD CHECK` note. We test all those functions in unit tests, so if we break _the functions themselves_  we'll know about it.

**Won't using `nrounds = 5L` and other parameter changes result in terrible models?**

Yes probably. But we don't have strong guarantees about how well the models from our example code work anyway.  I think the role of examples here is just to show minimal valid sets of arguments to each function. See  https://github.com/microsoft/LightGBM/issues/2988#issuecomment-611726510 for my full opinion on this.

### CODE REVIEW

**Comment by jameslamb** [comment]
@Laurae2 or @guolinke could you please take a look whenever you have time? This NOTE is blocking #2985 and even if we work around it there it's going to block other PRs (since the  runtime right now is so close to 5  seconds)

**Comment by StrikerRUS** [comment]
> Should we really be using \donttest? That means CI won't catch those examples being broken

Sorry for maybe stupid question, but I'm still far away from R and learning it only by our conversations here 😃 

As we do not wrap them into `\dontrun` (but in `\donttest`), it means that these examples are will be running at our RTD site and failing docs will indirectly indicate that our examples may be broken, right?

File: R-package/R/lgb.cv.R [hunk 1]
```diff
@@ -56,19 +56,17 @@ CVBooster <- R6::R6Class(
...
 #' model <- lgb.cv(
 #'   params = params
 #'   , data = dtrain
-#'   , nrounds = 10L
+#'   , nrounds = 5L
 #'   , nfold = 3L
 #'   , min_data = 1L
-#'   , learning_rate = 1.0
-#'   , early_stopping_rounds = 5L
+#'   , learning_rate = 0.05
```
Comment (Laurae2) [inline R-package/R/lgb.cv.R:69]: `learning_rate = 1`

File: R-package/R/lgb.plot.interpretation.R [hunk 2]
```diff
@@ -15,28 +15,43 @@
...
+#'   , label = labels
+#' )
+#' setinfo(dtrain, "init_score", rep(Logit(mean(labels)), length(labels)))
+#'
 #' data(agaricus.test, package = "lightgbm")
-#' test <- agaricus.test
 #'
 #' params <- list(
 #'   objective = "binary"
 #'   , learning_rate = 0.01
```
Comment (Laurae2) [inline R-package/R/lgb.plot.interpretation.R:34]: I would probably recommend `learning_rate = 0.1` there

File: R-package/R/lgb.plot.importance.R [hunk 3]
```diff
@@ -25,16 +25,18 @@
 #' params <- list(
 #'     objective = "binary"
 #'     , learning_rate = 0.01
```
Comment (Laurae2) [inline R-package/R/lgb.plot.importance.R:27]: I would probably recommend `learning_rate = 0.1` there

File: R-package/R/lgb.interprete.R [hunk 4]
```diff
@@ -28,12 +27,15 @@
 #' params <- list(
 #'     objective = "binary"
 #'     , learning_rate = 0.01
```
Comment (Laurae2) [inline R-package/R/lgb.interprete.R:29]: I would probably recommend `learning_rate = 0.1` there

File: R-package/R/lgb.importance.R [hunk 5]
```diff
@@ -13,20 +13,22 @@
...
 #'
 #' @examples
-#' library(lightgbm)
 #' data(agaricus.train, package = "lightgbm")
 #' train <- agaricus.train
 #' dtrain <- lgb.Dataset(train$data, label = train$label)
 #'
 #' params <- list(
 #'   objective = "binary"
 #'   , learning_rate = 0.01
```
Comment (Laurae2) [inline R-package/R/lgb.importance.R:22]: I would probably recommend `learning_rate = 0.1` there

**Comment by Laurae2** [comment]
> As we do not wrap them into `\dontrun` (but in `\donttest`), it means that these examples are will be running at our RTD site and failing docs will indirectly indicate that our examples may be broken, right?

@StrikerRUS The examples would fail silently (but still show the error in doc).

**Comment by jameslamb** [comment]
> > As we do not wrap them into `\dontrun` (but in `\donttest`), it means that these examples are will be running at our RTD site and failing docs will indirectly indicate that our examples may be broken, right?
> 
> @StrikerRUS The examples would fail silently (but still show the error in doc).

I thought that if you added `\dontrun`, it will literally add the comment "## don't run" in the examples which I think would be confusing. I swear it used to work like that. But I just tested and I guess not?

image

@Laurae2 is right (https://github.com/microsoft/LightGBM/pull/2989#issuecomment-613026665). I tried adding `stop("I am an error")` to the examples in `lgb.cv()` and then re-running `pkgdown::build_site()`.

**Results when that example is wrapped in `\donttest`**

image

**Results with `\dontrun`**

image

In which case I think the `\donttest` is still an ok choice. I'd rather that we have an error message that someone sees and alerts us to than just having the code be silently broken until someone runs it. I'd hate to have people run that code, get an error, and get discouraged and thing they're doing something wrong.

**Comment by jameslamb** [comment]
> Change the `learning_rate` to adapt to the hyperparameter changes

Thanks! Just made these changes

**Comment by StrikerRUS** [comment]
Not sure that the root cause is this PR, but we have the following right now at our docs:

image

https://lightgbm.readthedocs.io/en/latest/R/reference/lgb.cv.html

**Comment by StrikerRUS** [comment]
Also, it seems that two other examples are broken:

https://lightgbm.readthedocs.io/en/latest/R/reference/lgb.Dataset.set.categorical.html
https://lightgbm.readthedocs.io/en/latest/R/reference/lgb.plot.interpretation.html

**Comment by jameslamb** [comment]
Thanks @StrikerRUS ! I don't think they are related to this PR. I just created #2999 to address them though, so appreciate you pointing them out!

**Pull request merged**